components/batcher/batcher.go

```diff
@@ -79,43 +87,67 @@ func NewBatcher(parentCtx context.Context, cfg Config, l log.Logger, m metrics.M
 		return nil, err
 	}
 
-	ctx, cancel := context.WithCancel(parentCtx)
-
 	batchSubmitter, err := NewBatchSubmitter(cfg, l, m)
 	if err != nil {
-		cancel()
 		return nil, fmt.Errorf("failed to init batch submitter: %w", err)
 	}
 
-	balance, err := cfg.L1Client.BalanceAt(ctx, cfg.From, nil)
+	balance, err := cfg.L1Client.BalanceAt(parentCtx, cfg.TxManager.From(), nil)
 	if err != nil {
-		cancel()
 		return nil, err
 	}
 
-	l.Info("creating batcher", "batcher_addr", cfg.From, "batcher_bal", balance)
+	l.Info("creating batcher", "batcher_addr", cfg.TxManager.From(), "batcher_bal", balance)
 
 	return &Batcher{
-		ctx:            ctx,
-		cancel:         cancel,
 		cfg:            cfg,
 		l:              l,
 		batchSubmitter: batchSubmitter,
-		txMgr:          txmgr.NewSimpleTxManager("batcher", l, cfg.TxManagerConfig, cfg.L1Client),
+		txMgr:          cfg.TxManager,
 	}, nil
 }
 
 func (b *Batcher) Start() {
-	b.l.Info("starting Batch Submitter")
+	b.l.Info("starting Batcher")
+
+	b.shutdownCtx, b.cancelShutdownCtx = context.WithCancel(context.Background())
+	b.killCtx, b.cancelKillCtx = context.WithCancel(context.Background())
+
 	b.wg.Add(1)
 	go b.loop()
+
+	b.l.Info("Batcher started")
 }
 
-func (b *Batcher) Stop() {
-	b.cancel()
+func (b *Batcher) Stop(ctx context.Context) {
+	b.l.Info("Stopping Batcher")
+
+	// go routine will call cancelKillCtx() if the passed in ctx is ever Done
+	wrapped, cancel := context.WithCancel(ctx)
+	defer cancel()
+	go func() {
+		<-wrapped.Done()
+		b.cancelKillCtx()
```

Comment:
This part is slightly different from  https://github.com/ethereum-optimism/optimism/commit/99c75b47755b7d3b98baeb426d2a41459fa2a7ca

Why did you remove `cancleKill := b.cancelKillCtx`?
Plus, if `wrapped` isn't used anywhere, `<-wrapped.Done()` is called?

Reply:
- The reason I removed `cancleKill := b.cancelKillCtx` is because that function is used twice in line 130 and 135, so I wanted to use it in a unified form. Also I wanted to use it in a unified form with `b.cancelShutdownCtx()`.
- I thought `<-wrapped.Done()` will be called when `defer cancel()` is called in line 127. So I added the goroutine in line 128 as a sort of safeguard to cancel `b.killCtx` in any case. Please let me know if I understand it incorrectly.

Reply:
Changed to declare `cancleKill := b.cancelKillCtx` and use it.